File: electives/JPP/L3/cpp/lib/GF.cpp
//
// Created by jaskow on 06.04.24.
//
#include <stdexcept>
#include "GF.hpp"
#include "lib.h"

//////////////////////////////////////////////////////////////
// private
void GF::checkP(const GF& L, const GF& R) {
    if(L.p != R.p)
        throw std::invalid_argument("[GF]ERROR: not matching p! " + std::to_string(L.p) + " != " + std::to_string(R.p));
}

uint32_t GF::inverse(const uint32_t _p, const int64_t _val){
    return _p + (_val % _p);
}


//////////////////////////////////////////////////////////////
// public
GF::GF(const uint32_t& _p, const int64_t& _val) {
    if(_p == 0) {
        throw std::invalid_argument("[GF]ERROR: p must be > 0");
    }
    p = _p;
    setVal(_val);
}


uint32_t GF::getP() const noexcept{
    return p;
}

void GF::setP(uint32_t _p) noexcept{
    p = _p;
    this->setVal(val);
}

uint32_t GF::getVal() const noexcept{
    return val;
}

void GF::setVal(int64_t _val) noexcept{
    if(_val < 0 ) {
        val = inverse(p, _val);
    }
    else {
        val = _val % p;
    }
}


// compare
bool GF::operator==(const GF &R) const {
    return this->val == R.val;
}
bool GF::operator!=(const GF &R) const {
    return !(*this == R);
}
bool GF::operator<(const GF &R) const {
    checkP(*this, R);
    return this->val < R.val;
}
bool GF::operator>(const GF &R) const {
    return R < *this;
}
bool GF::operator<=(const GF &R) const {
    return !(*this > R);
}
bool GF::operator>=(const GF &R) const {
    return !(*this < R);
}


// arithmetic
GF GF::operator+() const noexcept {
    return *this;
}
GF GF::operator-() const noexcept{
    return GF(p, p - val);
}
GF GF::operator-(const GF &R) const {
    checkP(*this, R);
    return ((-R) + (*this));
}
GF GF::operator+(const GF &R) const {
    checkP(*this, R);
    return GF(R.p, ((uint64_t) val + R.val) % p);
}
GF GF::operator*(const GF &R) const {
    checkP(*this, R);
    return GF(p, ((uint64_t) val * R.val) % p); 
}
GF GF::operator/(const GF &R) const {
    checkP(*this, R);

    uint64_t RValReverse = ILDES(R.val, R.p, RGCD(R.val, R.p)).x;

    return GF(p, val) * GF(p, RValReverse);
}


// assignment
GF& GF::operator+=(const GF &other) {
    *this = ((*this) + other);
    return *this;
}
GF& GF::operator-=(const GF &other) {
    *this = (*this - other);
    return *this;
}
GF& GF::operator*=(const GF &other) {
    *this = ((*this) * other);
    return *this;
}
GF& GF::operator/=(const GF &other) {
    *this = (*this / other);
    return *this;
}


std::ostream &operator<<(std::ostream &out, const GF &gf) {
    out << "p: " << gf.p << ", "
        << "val: " << gf.val;
    return out;
}

File: electives/JPP/L3/java/src/main/java/org/example/GF.java
package org.example;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

public class GF implements GFI {
    private long p;
    private long val;

    private static void checkP(@NotNull GF L, @NotNull GF R) {
        if(L.p != R.p)
            throw new IllegalArgumentException("[GF]ERROR: not matching p!");
    }
    private static long inverse(long _p, long _val){
        return _p + (_val % _p);
    }

    public GF() {
        p = 0;
        setVal(0);
    }
    public GF(long _p, long _val) {
        if(_p == 0) {
            throw new IllegalArgumentException("[GF]ERROR: p must be > 0");
        }
        p = _p;
        setVal(_val);
    }

    @Override
    public GF instantiate(long _p, long _val) {
        return new GF();
    }
    @Override
    public long getP() { return p; }
    @Override
    public void setP(long _p) { p = _p; setVal(val); }
    @Override
    public long getVal() {
        return val;
    }
    @Override
    public void setVal(long _val) {
        if(_val < 0 ) {
            val = inverse(p, _val);
        }
        else {
            val = _val % p;
        }
    }


    // compare
    @Override
    public boolean equals(@NotNull GF L, @NotNull GF R) {
        checkP(L, R);
        return L.val == R.val;
    }
    @Override
    public boolean notEquals(@NotNull GF L, @NotNull GF R) {
        return !equals(L, R);
    }
    @Override
    public boolean lessThan(@NotNull GF L, @NotNull GF R) {
        checkP(L, R);
        return L.val < R.val;
    }
    @Override
    public boolean greaterThan(@NotNull GF L, @NotNull GF R) {
        return lessThan(R, L);
    }
    @Override
    public boolean lessThanOrEqual(@NotNull GF L, @NotNull GF R) {
        return !greaterThan(L, R);
    }
    @Override
    public boolean greaterThanOrEqual(@NotNull GF L, @NotNull GF R) {
        return !lessThan(L, R);
    }

    // arithmetic
    @Override
    public GF plus() {
        return this;
    }
    @Override
    public GF minus() {
        return new GF(this.p, this.p - this.val);
    }
    @Override
    @Contract("_, _ -> new")
    public GF add(@NotNull GF L, @NotNull GF R) {
        checkP(L, R);
        long _val = (L.val + R.val) % L.p;
        return new GF(R.p, _val);
    }
    @Override
    @Contract("_, _ -> new")
    public GF subtract(@NotNull GF L, @NotNull GF R) {
        checkP(L, R);
        R = R.minus();
        return add(R, L);
    }
    @Override
    @Contract("_, _ -> new")
    public GF multiply(@NotNull GF L, @NotNull GF R) {
        checkP(L, R);
        return new GF(L.p, (L.val * R.val) % L.p);
    }
    @Override
    @Contract("_, _ -> new")
    public GF divide(@NotNull GF L, @NotNull GF R) {
        checkP(L, R);
        long RValReverse = Lib.ILDES(R.val, R.p, Lib.IGCD(R.val, R.p))[0];
        GF result = new GF(R.p, RValReverse);
        return multiply(L, result);
    }

    // assign
    @Override
    public void assign(@NotNull GF other) {
        setP(other.p);
        setVal(other.val);
    }
    @Override
    public void plusAssign(@NotNull GF other) {
        val = add(this, other).val;
    }
    @Override
    public void minusAssign(@NotNull GF other) {
        val = subtract(this, other).val;
    }
    @Override
    public void multiplyAssign(@NotNull GF other) {
        val = multiply(this, other).val;
    }
    @Override
    public void divideAssign(@NotNull GF other) {
        val = divide(this, other).val;
    }

    @Override
    public String toString() {
        return "p   = " + p + "\nval = " + val;
    }
}
